File: SafetyProgram.DocumentObjects/ChemicalTableNs/Ribbon/ChemicalTableRibbonTabViewModel.cs
using System.Collections.ObjectModel;
using System.ComponentModel;
using SafetyProgram.Base;
using SafetyProgram.Base.Interfaces;
using SafetyProgram.DocumentObjects.ChemicalTableNs.Commands;

namespace SafetyProgram.DocumentObjects.ChemicalTableNs.Ribbon
{
    internal sealed class ChemicalTableRibbonTabViewModel : IChemicalTableRibbonViewModel
    {
        private readonly IConfiguration appConfiguration;

        public ChemicalTableRibbonTabViewModel(IConfiguration appConfiguration,
            IChemicalTableCommands commands)
        {
            Helpers.NullCheck(appConfiguration, commands);

            this.appConfiguration = appConfiguration;
            this.commands = commands;

            //Perform an initial blank search
            performSearch("");
        }

        private string search = "";
        public string Search
        {
            get
            {
                return search;
            }
            set
            {
                search = value;
                PropertyChanged.Raise(this, "Search");
                performSearch(search);
            }
        }

        private readonly ObservableCollection<IChemicalModelObject> searchResult = new ObservableCollection<IChemicalModelObject>();
        public ObservableCollection<IChemicalModelObject> SearchResult
        {
            get { return searchResult; }
        }

        private readonly IChemicalTableCommands commands;
        public IChemicalTableCommands Commands
        {
            get { return commands; }
        }

        private void performSearch(string searchString)
        {
            searchResult.Clear();

            if (searchString == "")
            {
                foreach (IRepository<IChemicalModelObject> repository in appConfiguration.ChemicalRepositories)
                {
                    repository.GetAll(
                        ((retrievedChemical) => searchResult.Add(retrievedChemical))
                    );
                }
            }
            else
            {
                foreach (IRepository<IChemicalModelObject> repository in appConfiguration.ChemicalRepositories)
                {
                    repository.Get(
                        ((someChemical) =>
                            someChemical.Name.ToLower().Contains(searchString.ToLower())
                        ),
                        (foundChemical) => searchResult.Add(foundChemical)
                    );
                }
            }
        }

        public event PropertyChangedEventHandler PropertyChanged;
    }
}

File: SafetyProgram.UI.ViewModels/ChemicalTableRibbonTabViewModel.fs
namespace SafetyProgram.UI.ViewModels

open System.Collections.ObjectModel
open SafetyProgram.UI.Models

type ChemicalTableRibbonTabViewModel(model : GuiChemicalTable) = 

    let propertyChanged = new Event<_,_>()
    let mutable search = ""
    let searchResult = new ObservableCollection<_>()

    interface IChemicalTableRibbonTabViewModel with

        member this.Search 
            with get () = search
            and set x = search <- x

        member this.SearchResult = searchResult

        [<CLIEvent>]
        member this.PropertyChanged = propertyChanged.Publish

        member this.Dispose() = ()

    // IMPLICIT IMPL
    member this.Search 
            with get () = search
            and set x = search <- x

    member this.SearchResult = searchResult

    [<CLIEvent>]
    member this.PropertyChanged = propertyChanged.Publish

    member this.Dispose() = ()